File: buildingowners.lua
PLUGIN.Name = "buildingowners"
PLUGIN.Title = "Building Owners"
PLUGIN.Version = V(1, 3, 2)
PLUGIN.Description = "Manage building owners"
PLUGIN.Author = "Reneb"
PLUGIN.HasConfig = true
PLUGIN.ResourceId = 682

function PLUGIN:Init()
	self.ServerInitialized = false
	command.AddChatCommand( "changeowner", self.Object, "cmdChangeOwner" )
	timer.Once( 0.1, function() 
		self.ServerInitialized = true 
		
	end )
	self:LoadSavedData()
end
----------- All local functions for this plugin only -----------

----------- Get the position of the ground -----------
local function GetGround(pos)
	local Ray = new( UnityEngine.Ray._type, nil )
	Ray.direction = UnityEngine.Vector3.get_down()
	Ray.origin = pos
	local arr = util.TableToArray( { Ray } )
	local hits = UnityEngine.Physics.RaycastAll["methodarray"][1]:Invoke(nil,arr)
	local it = hits:GetEnumerator()
	local gotinfo = {}
	while (it:MoveNext()) do
		if(tostring(it.Current.collider.gameObject.name) == "Terrain") then
			return it.Current.point
		end
	end
	return false
end

----------- Get any building block around an entity -----------
local buildingradius = 3
local function FindBuilding(entity)
	local arr = util.TableToArray( { entity.transform.position , buildingradius } )
	util.ConvertAndSetOnArray(arr, 1, buildingradius, System.Single._type)
	local hits = UnityEngine.Physics.OverlapSphere["methodarray"][1]:Invoke(nil,arr)
	local it = hits:GetEnumerator()
	local buildingblock = false
	local distance = 9999
	while (it:MoveNext()) do
		if(it.Current:GetComponentInParent(global.BuildingBlock._type)) then
			if(string.find(tostring(it.Current:GetComponentInParent(global.BuildingBlock._type).blockDefinition.name),"foundation")) then
				buildingblock =  it.Current:GetComponentInParent(global.BuildingBlock._type)
				break
			end
		end
	end
	return buildingblock
end

----------- Load and save building owners data -----------
function PLUGIN:SaveData()  
    datafile.SaveDataTable( "BuildingOwners" )
end
function PLUGIN:LoadSavedData()
    ReverseData = datafile.GetDataTable( "BuildingOwners" )
    ReverseData = ReverseData or {}
	self:ReverseTable()
end

function PLUGIN:ReverseTable()
	OwnersData = {}
	for steamid,datable in pairs(ReverseData) do
		for dataid,height in pairs(datable) do
			OwnersData[height] = steamid
		end
	end
	print("Successfully initialized Building Owners")
end

----------- Load default configs -----------
function PLUGIN:LoadDefaultConfig()
	self.Config.useByType = 1
	self.Config.ChangeOwnerForModerators = true
end

----------- Change owner command -----------
function PLUGIN:cmdChangeOwner( player, com, args )
	local authlevel = player:GetComponent("BaseNetworkable").net.connection.authLevel
	local neededlevel = 2
	if(self.Config.ChangeOwnerForModerators) then
		neededlevel = 1
	end
	if(authlevel and authlevel >= neededlevel) then
		if(args.Length >= 1) then
			----------- Get a player from the built in rust way -----------
			local targetplayer = global.BasePlayer.Find(tostring(args[0]))
			if(not targetplayer) then
				----------- Get a player from my way -----------
				local plistenum = player.activePlayerList
				local it = plistenum:GetEnumerator()
				while (it:MoveNext()) do
					if(string.find(it.Current.displayName,tostring(args[0]))) then
						if(targetplayer) then
							rust.SendChatMessage(player,"SERVER","Multiple players found")
							return
						end
						targetplayer = it.Current
					end
				end
				if(not targetplayer) then
					rust.SendChatMessage(player,"SERVER","This player " .. args[0] .. " doesn't exist")
					return
				end
			end
			local house = {}
			local gotdeployeditem = {}
			local checkfrom = {}
			local current = 0
			local buildingblock = FindBuilding(player)
			if(not buildingblock) then
				rust.SendChatMessage(player,"SERVER","You must be standing on a foundation of the building that you want to change the owner of")
				return
			end
			local posy = buildingblock.transform.position.y
			local userid = rust.UserIDFromPlayer(targetplayer)
			OwnersData[tostring(posy)] = userid
			
			if(not ReverseData[userid]) then ReverseData[userid] = {} end
			table.insert(ReverseData[userid],tostring(posy))
			local userid = rust.UserIDFromPlayer(targetplayer)
			local house = {}
			local checkfrom = {}
			local radius = 3
			house[buildingblock] = true
			checkfrom[ #checkfrom + 1 ] = buildingblock.transform.position
			local current = 0
			local gotdeployeditem = {}
			while(true) do
				current = current + 1
				if(not checkfrom[current]) then
					break
				end
				local arr = util.TableToArray( { checkfrom[current] , radius } )
				util.ConvertAndSetOnArray(arr, 1, radius, System.Single._type)
				local hits = UnityEngine.Physics.OverlapSphere["methodarray"][1]:Invoke(nil,arr)
				local it = hits:GetEnumerator()
				while (it:MoveNext()) do
					if(it.Current:GetComponentInParent(global.BuildingBlock._type)) then
						local fbuildingblock = it.Current:GetComponentInParent(global.BuildingBlock._type)
						if(not house[fbuildingblock]) then
							house[fbuildingblock] = true
							local posy = fbuildingblock.transform.position.y
							local groundpos = GetGround(fbuildingblock.transform.position)
							if(not groundpos) then groundpos = fbuildingblock.transform.position end
							local levels = math.floor( (posy - groundpos.y) / 3)
							local theposy = posy - (levels * 3)
							checkfrom[ #checkfrom + 1 ] = fbuildingblock.transform.position
							OwnersData[tostring(theposy)] = userid
							table.insert(ReverseData[userid],tostring(theposy))							
						end
					end
				end
			end
			rust.SendChatMessage(player,"SERVER","New owner of this house has been set to: " .. targetplayer.displayName)
			self:SaveData()
		end
	end
end

----------- Hook that sets a building owner for every buildingblock built (if needed) -----------
function PLUGIN:OnEntityBuilt(helditem, gameobject)
	if(self.ServerInitialized) then
		local buildingblock = gameobject:GetComponent("BuildingBlock")
		local posy = buildingblock.transform.position.y
		if(not OwnersData[tostring(posy)]) then
			local userid = rust.UserIDFromPlayer(helditem.ownerPlayer)
			OwnersData[tostring(posy)] = userid
			if(not ReverseData[userid]) then ReverseData[userid] = {} end
			table.insert(ReverseData[userid],tostring(posy))		
			self:SaveData()
		end
	end
end
function PLUGIN:AddBlockData(buildingblock,baseplayer)
	local userID = rust.UserIDFromPlayer(baseplayer)
	local pos = buildingblock.transform.position
	OwnersData[tostring(pos.y)] = userID
	if(not ReverseData[userID]) then ReverseData[userID] = {} end
	table.insert(ReverseData[userID],tostring(pos.y))		
end

----------- Function to get the owner of a building block from another plugins. -----------
function PLUGIN:FindBlockData(buildingblock)
	local posy = buildingblock.transform.position.y
	local groundpos = GetGround(buildingblock.transform.position)
	if(not groundpos) then groundpos = buildingblock.transform.position end
	local levels = math.floor( (posy - groundpos.y) / 3)
	local theposy = posy - (levels * 3)
	if(OwnersData[tostring(theposy)]) then
		return OwnersData[tostring(theposy)]
	end
	return false
end

File: CSharp/BuildingOwners.cs
using System.Collections.Generic;
using System;
using System.Reflection;
using System.Data;
using UnityEngine;
using Oxide.Core;

namespace Oxide.Plugins
{
    [Info("Building Owners", "Reneb", "2.1.3", ResourceId = 682)]
    class BuildingOwners : RustPlugin
    {
        private static DateTime epoch;
        private Core.Configuration.DynamicConfigFile ReverseData;
        private Dictionary<float,string> OwnersData;
        private FieldInfo keyvalues;
        private static bool serverInitialized;
        void Loaded()
        {
            serverInitialized = false;
            OwnersData = new Dictionary<float, string>();
            epoch = new System.DateTime(1970, 1, 1);
            keyvalues = typeof(Oxide.Core.Configuration.DynamicConfigFile).GetField("keyvalues", (BindingFlags.Public | BindingFlags.Static | BindingFlags.Instance | BindingFlags.NonPublic));
        }
        void OnServerInitialized()
        {
            ReverseData = Interface.GetMod().DataFileSystem.GetDatafile("BuildingOwners");
            serverInitialized = true;
            ReverseTable();
        }
        void SaveData()
        {
            Interface.GetMod().DataFileSystem.SaveDatafile("BuildingOwners");
        }
        void ReverseTable()
        {
            //var Table = keyvalues.GetValue(ReverseData) as Dictionary<string, object>;

            foreach (KeyValuePair<string, object> pair in ReverseData)
            {
                var list = pair.Value as List<object>;
                foreach (object heights in list)
                {
                    if (!(OwnersData.ContainsKey(Convert.ToSingle(heights))))
                    {
                        OwnersData.Add(Convert.ToSingle(heights), pair.Key.ToString());
                    }
                }
            }
        }
        double CurrentTime()
        {
            return System.DateTime.UtcNow.Subtract(epoch).TotalMilliseconds;
        }
        object FindBuilding(BasePlayer player)
        {
            var hits = UnityEngine.Physics.OverlapSphere(player.transform.position, 3f);
            foreach (var hit in hits)
            {
                if (hit.GetComponentInParent<BuildingBlock>() != null)
                {
                    return hit.GetComponentInParent<BuildingBlock>();
                }
            }
            return false;
        }
        void OnEntityBuilt(HeldEntity heldentity, GameObject gameobject)
        {
            if (serverInitialized)
            {
                var buildingblock = gameobject.GetComponent<BuildingBlock>();
                if (buildingblock == null) return;
                float posy = buildingblock.transform.position.y;
                if (!(OwnersData.ContainsKey(posy)))
                {
                    string userid = heldentity.ownerPlayer.userID.ToString();
                    OwnersData.Add(posy, userid);
                    if (ReverseData[userid] == null)
                        ReverseData[userid] = new List<object>();
                    var list = ReverseData[userid] as List<object>;
                    list.Add(posy);
                    ReverseData[userid] = list;
                }
            }
        }
        object FindBlockData(BuildingBlock block)
        {
            float posy = block.transform.position.y;
            if (OwnersData.ContainsKey(posy))
                return OwnersData[posy];
            return false;
        }
        [ChatCommand("changeowner")]
        void cmdChatchangeowner(BasePlayer player, string command, string[] args)
        {
            if (player.net.connection.authLevel < 1)
            {
                SendReply(player, "You are not allowed to use this command");
                return;
            }
            if (args == null || args.Length < 1)
            {
                SendReply(player, "You need to give the name of the new owner");
                return;
            }
            var target = BasePlayer.Find(args[0].ToString());
            if (target == null || target.net == null || target.net.connection == null)
            {
                SendReply(player, "Target player not found");
            }
            else
            {
                object block = FindBuilding(player);
                if ( block is bool )
                {
                    SendReply(player, "No Building found.");
                }
                else
                {
                    var userid = target.userID.ToString();
                    BuildingBlock buildingblock = block as BuildingBlock;
                    List<BuildingBlock> houseList = new List<BuildingBlock>();
                    List<Vector3> checkFrom = new List<Vector3>();
                    houseList.Add((BuildingBlock) buildingblock);
                    checkFrom.Add(buildingblock.transform.position);
                    var current = 0;
                    while (true)
                    {
                        current++;
                        if (current > checkFrom.Count)
                            break;
                        var hits = UnityEngine.Physics.OverlapSphere(checkFrom[current-1], 3f);
                        foreach (var hit in hits)
                        {
                            if (hit.GetComponentInParent<BuildingBlock>() != null)
                            {
                                BuildingBlock fbuildingblock = hit.GetComponentInParent<BuildingBlock>();
                                if (!(houseList.Contains(fbuildingblock)))
                                {
                                    houseList.Add(fbuildingblock);
                                    checkFrom.Add(fbuildingblock.transform.position);
                                    if (!(OwnersData.ContainsKey(fbuildingblock.transform.position.y)))
                                    {
                                        OwnersData.Add(fbuildingblock.transform.position.y, userid);
                                    }
                                    else
                                    {
                                        OwnersData[fbuildingblock.transform.position.y] = userid;
                                    }
                                    if (ReverseData[userid] == null)
                                        ReverseData[userid] = new List<object>();
                                    var list = ReverseData[userid] as List<object>;
                                    if (!(list.Contains(fbuildingblock.transform.position.y)))
                                    {
                                        list.Add(fbuildingblock.transform.position.y);
                                        ReverseData[userid] = list;
                                    }
                                }
                            }
                        }
                    }
                    SendReply(player, string.Format("New owner of this house is: {0}",target.displayName));
                    SendReply(target, "An admin gave you the ownership of this house");
                }
            }
        }
        void OnServerSave()
        {
            SaveData();
        }
        void OnServerQuit()
        {
            SaveData();
        }
    }
}
